### PR TITLE
Update WebBluetooth to use Promises

### PR DESCRIPTION
**Pull request by dati91**
Initial patch to support promises in WebBluetooth.
This will allow us to use the expected syntax in webpages.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13428)

### CODE REVIEW

File: components/script/dom/bluetooth.rs [hunk 1]
```diff
@@ -252,6 +272,31 @@ fn canonicalize_filter(filter: &BluetoothRequestDeviceFilter, global: GlobalRef)
                                 service_data_uuid))
 }
 
+#[allow(unrooted_must_root)]
+#[allow(unsafe_code)]
+pub fn result_to_promise<T: ToJSValConvertible>(global_ref: GlobalRef,
+                                                 bluetooth_result: Fallible<T>)
+                                                 -> Fallible<Rc<Promise>> {
+    match bluetooth_result {
```
Comment (jdm) [inline components/script/dom/bluetooth.rs]:
```rust
let p = Promise::new(global_ref);
match bluetooth_result {
    Ok(v) => p.resolve_native(v),
    Err(e) => p.reject_error(e),
}
```

File: components/script/dom/bluetoothremotegattcharacteristic.rs [hunk 2]
```diff
@@ -213,3 +194,55 @@ impl BluetoothRemoteGATTCharacteristicMethods for BluetoothRemoteGATTCharacteris
...
+        self.service.get()
+    }
+
+    // https://webbluetoothcg.github.io/web-bluetooth/#dom-bluetoothremotegattcharacteristic-uuid
+    fn Uuid(&self) -> DOMString {
+        self.uuid.clone()
+    }
+
+    #[allow(unrooted_must_root)]
+    #[allow(unsafe_code)]
```
Comment (jdm) [inline components/script/dom/bluetoothremotegattcharacteristic.rs]: Don't need this and similar ones in this file and other files.

File: components/script/dom/bluetooth.rs [hunk 3]
```diff
@@ -252,6 +272,31 @@ fn canonicalize_filter(filter: &BluetoothRequestDeviceFilter, global: GlobalRef)
                                 service_data_uuid))
 }
 
+#[allow(unrooted_must_root)]
+#[allow(unsafe_code)]
+pub fn result_to_promise<T: ToJSValConvertible>(global_ref: GlobalRef,
+                                                 bluetooth_result: Fallible<T>)
+                                                 -> Fallible<Rc<Promise>> {
```
Comment (jdm) [inline components/script/dom/bluetooth.rs]: nit: indentation

File: components/script/dom/webidls/Bluetooth.webidl [hunk 4]
```diff
@@ -23,9 +23,8 @@ interface Bluetooth {
 //  [SecureContext]
 //  readonly attribute BluetoothDevice? referringDevice;
 //  [SecureContext]
-//  Promise<BluetoothDevice> requestDevice(RequestDeviceOptions options);
   [Throws]
```
Comment (jdm) [inline components/script/dom/webidls/Bluetooth.webidl]: This doesn't need to be here anymore, or most (all?) of the other ones in other bluetooth webidl files.

**Review by jdm — CHANGES_REQUESTED** [review]
Great improvement!

File: components/script/dom/webidls/Bluetooth.webidl (hunk 4 above)
Comment (dati91) [inline components/script/dom/webidls/Bluetooth.webidl]: Do you mean we should remove the Throws/Fallible from the return types?

Comment (jdm) [inline components/script/dom/webidls/Bluetooth.webidl]: Specifically, the [Throws] annotation, since none of the methods throw an exception any more.

**Comment by jdm** [comment]
@bors-servo: delegate+
Just say `@bors-servo: r=jdm` after squashing the commits into one. Thanks!

**Comment by jdm** [comment]
@bors-servo: retry
* #12654

**Pull request merged**